File: 238.py
class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        n = len(nums)
        result = [1] * n
        left = 1
        right = 1
        for i in range(n):
            result[i] *= left
            result[n - i - 1] *= right
            left *= nums[i]
            right *= nums[n - i - 1]
        return result
    
if __name__ == '__main__':
    s = Solution()
    nums = list(map(int, input().split()))
    print(s.productExceptSelf(nums))

File: cpp/238.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums){
        int n = nums.size();
        vector<int> ans(n, 1);
        int left = 1, right = 1;
        for(int i = 0; i < n; i++){
            ans[i] *= left;
            left *= nums[i];
            ans[n - 1 - i] *= right;
            right *= nums[n - 1 - i];
        }
        return ans;
    }
};

int main(){
    Solution s;
    int n;
    cin >> n;
    vector<int> nums(n);
    for(int i = 0; i < n; i++){
        cin >> nums[i];
    }
    vector<int> res = s.productExceptSelf(nums);
    for(int i = 0; i < n; i++){
        cout << res[i] << " ";
    }
    cout << endl;
    return 0;
}